core/service.go

```diff
@@ -54,10 +55,24 @@ type Config struct {
 // NewService returns a new core service that connects the runtime, BABE
 // session, and p2p service.
 func NewService(cfg *Config, newBlocks chan types.Block) (*Service, error) {
+	if cfg.Keystore == nil {
+		return nil, fmt.Errorf("no keystore provided")
+	}
+
+	keys := cfg.Keystore.Sr25519Keypairs()
+	// no validator keypair found, generate a new one
+	if len(keys) == 0 {
+		kp, err := sr25519.GenerateKeypair()
+		if err != nil {
+			return nil, err
+		}
+		cfg.Keystore.Insert(kp)
+		keys = cfg.Keystore.Sr25519Keypairs()
+	}
 
 	// BABE session configuration
 	bsConfig := &babe.SessionConfig{
-		Keystore:  cfg.Keystore,
+		Keypair:   keys[0].(*sr25519.Keypair),
```

Comment:
This seems very unsafe

Reply:
yeah.. I mean if the keystore is empty, we generate a new keypair, so keys[0] should always exist. there probably is a better way to specify what key we want, but we might want to use different keys for different babe sessions, which makes it a bit more complex.